File: java/src/main/java/com/eluke/checkers/KingPiece.java
package com.eluke.checkers;

import java.util.Arrays;
import java.util.Collection;

public class KingPiece extends Piece {

	public KingPiece(final Piece piece) {
		super(piece.color());
	}

	public KingPiece(final Color color) {
		super(color);
	}

	@Override
	public Collection<Move> getSimpleMoves(final Board board, final Coordinate coordinate) {
		return MoveRules.getSimpleMoves(board, this, coordinate, this::getFollowingCoords);
	}

	@Override
	public Collection<Move> getJumpMoves(final Board board, final Coordinate coordinate) {
		return MoveRules.getJumpMoves(board, this, coordinate, this::getFollowingCoords);
	}

	@Override
	public boolean canBeKinged() {
		return false;
	}

	private Collection<Coordinate> getFollowingCoords(final Coordinate c) {
		return Arrays.asList(
				new Coordinate(c.getRow()+1, c.getCol()+1),
				new Coordinate(c.getRow()+1, c.getCol()-1),
				new Coordinate(c.getRow()-1, c.getCol()+1),
				new Coordinate(c.getRow()-1, c.getCol()-1));
	}

	@Override
	public String toString() {
		return this.color() + " King";
	}
}

File: cpp/Checkers-CPP/KingPiece.cpp
/*
 * KingPiece.cpp
 *
 *  Created on: Mar 25, 2017
 *      Author: luke
 */

#include <functional>
#include "KingPiece.h"
#include "Coordinate.h"
#include "MoveRules.h"

std::list<Move*> KingPiece::getSimpleMoves(Board& board,
		const Coordinate& coordinate) const {

	return MoveRules::getSimpleMoves(board, this, coordinate, std::mem_fn (&KingPiece::getCoordinates));
}

std::list<Move*> KingPiece::getJumpMoves(Board& board,
		const Coordinate& coordinate) const {
	return MoveRules::getJumpMoves(board, this, coordinate, std::mem_fn (&KingPiece::getCoordinates));
}


std::list<Coordinate> KingPiece::getCoordinates(
		const Coordinate& coordinate) const {
	std::list<Coordinate> coords;

	int row = coordinate.getRow();
	coords.push_back(Coordinate(row+1, coordinate.getCol()+1));
	coords.push_back(Coordinate(row+1, coordinate.getCol()-1));
	coords.push_back(Coordinate(row-1, coordinate.getCol()+1));
	coords.push_back(Coordinate(row-1, coordinate.getCol()-1));
	return coords;
}
